File: src/viewer/assets/lib/units.js
/**
 * Unit formatting system for chart axes
 * Supports automatic scaling of values based on magnitude
 */

// Define unit scaling systems
const UNIT_SYSTEMS = {
  // Time-based units (nanoseconds to seconds)
  time: {
    base: 'ns',
    scales: [
      { threshold: 0, suffix: 'ns', divisor: 1 },
      { threshold: 1000, suffix: 'μs', divisor: 1000 },
      { threshold: 1000000, suffix: 'ms', divisor: 1000000 },
      { threshold: 1000000000, suffix: 's', divisor: 1000000000 }
    ]
  },
  
  // Data size (bytes)
  bytes: {
    base: 'B',
    scales: [
      { threshold: 0, suffix: 'B', divisor: 1 },
      { threshold: 1024, suffix: 'KB', divisor: 1024 },
      { threshold: 1048576, suffix: 'MB', divisor: 1048576 },
      { threshold: 1073741824, suffix: 'GB', divisor: 1073741824 },
      { threshold: 1099511627776, suffix: 'TB', divisor: 1099511627776 }
    ]
  },
  
  // Network data rate (bits per second)
  bitrate: {
    base: 'bps',
    scales: [
      { threshold: 0, suffix: 'bps', divisor: 1 },
      { threshold: 1000, suffix: 'Kbps', divisor: 1000 },
      { threshold: 1000000, suffix: 'Mbps', divisor: 1000000 },
      { threshold: 1000000000, suffix: 'Gbps', divisor: 1000000000 },
      { threshold: 1000000000000, suffix: 'Tbps', divisor: 1000000000000 }
    ]
  },
  
  // Percentage (already formatted, just add %)
  percentage: {
    base: '%',
    scales: [
      { threshold: 0, suffix: '%', divisor: 1, multiplier: 100 } // Added multiplier for percentage
    ]
  },

  // Frequency (Hz to GHz)
  frequency: {
    base: 'Hz',
    scales: [
      { threshold: 0, suffix: 'Hz', divisor: 1 },
      { threshold: 1000, suffix: 'KHz', divisor: 1000 },
      { threshold: 1000000, suffix: 'MHz', divisor: 1000000 },
      { threshold: 1000000000, suffix: 'GHz', divisor: 1000000000 }
    ]
  },
  
  // Count (no units, just numbers with K, M, B suffixes)
  count: {
    base: '',
    scales: [
      { threshold: 0, suffix: '', divisor: 1 },
      { threshold: 1000, suffix: 'K', divisor: 1000 },
      { threshold: 1000000, suffix: 'M', divisor: 1000000 },
      { threshold: 1000000000, suffix: 'B', divisor: 1000000000 }
    ]
  }
};

/**
 * Format a value with the appropriate unit scaling
 * 
 * @param {number} value - The value to format
 * @param {string} unitSystem - The unit system to use ('time', 'bytes', 'bitrate', etc.)
 * @param {number} precision - Number of decimal places to include (default: 2)
 * @return {object} Object with formatted value and unit string
 */
function formatWithUnit(value, unitSystem, precision = 2) {
  // Handle invalid or zero values
  if (value === null || value === undefined || isNaN(value)) {
    return { value: '0', unit: UNIT_SYSTEMS[unitSystem]?.base || '' };
  }
  
  // Get absolute value for scaling (we'll preserve sign later)
  const absValue = Math.abs(value);
  
  // Normalize unit system name - handle both 'time' and 'time_ns'
  const normalizedUnitSystem = unitSystem === 'time_ns' ? 'time' : unitSystem;
  
  // Get the unit system configuration
  const system = UNIT_SYSTEMS[normalizedUnitSystem];
  if (!system) {
    // Fallback for unknown unit systems
    return { 
      value: value.toFixed(precision), 
      unit: unitSystem || '' 
    };
  }
  
  // Find the appropriate scale for this value
  let scale = system.scales[0]; // Default to the smallest scale
  
  // Start from the largest scale and work backwards
  for (let i = system.scales.length - 1; i >= 0; i--) {
    if (absValue >= system.scales[i].threshold) {
      scale = system.scales[i];
      break;
    }
  }
  
  // Apply multiplier if needed (e.g., for percentages)
  const multiplier = scale.multiplier || 1;
  
  // Format the value with the selected scale and multiplier
  const scaledValue = ((value * multiplier) / scale.divisor).toFixed(precision);
  
  // Remove trailing zeros after decimal point
  const cleanValue = scaledValue.replace(/\.0+$/, '').replace(/(\.\d*[1-9])0+$/, '$1');
  
  return {
    value: cleanValue,
    unit: scale.suffix
  };
}

/**
 * Creates a formatter function for ECharts that applies unit scaling
 * 
 * @param {string} unitSystem - The unit system to use
 * @param {number} precision - Number of decimal places (default: 2)
 * @return {Function} Formatter function for ECharts
 */
function createAxisLabelFormatter(unitSystem, precision = 2) {
  return function(value) {
    // Skip formatting for empty values
    if (value === '' || value === null || value === undefined) return '';
    
    const formatted = formatWithUnit(value, unitSystem, precision);
    return formatted.value + (formatted.unit ? ' ' + formatted.unit : '');
  };
}


/**
 * Creates a tooltip formatter with appropriate unit scaling
 * 
 * @param {string} unitSystem - The unit system to use
 * @param {number} precision - Number of decimal places (default: 2)
 * @return {Function} Formatter function for tooltip values
 */
function createTooltipFormatter(unitSystem, precision = 2) {
  return function(params) {
    // The params object includes the value to format
    if (Array.isArray(params.value)) {
      // Handle scatter plots with [x, y] values
      // params.value[0] is the timestamp (already formatted)
      // params.value[1] is the actual value that needs unit formatting
      const formatted = formatWithUnit(params.value[1], unitSystem, precision);
      return `${formatted.value} ${formatted.unit}`;
    } else {
      // Handle normal line plots where params.value is the numeric value
      const formatted = formatWithUnit(params.value, unitSystem, precision);
      return `${formatted.value} ${formatted.unit}`;
    }
  };
}

// Export the utility functions
export {
  UNIT_SYSTEMS,
  formatWithUnit,
  createAxisLabelFormatter,
  createTooltipFormatter
};

File: src/viewer/tui/units.rs
//! Value formatting for TUI chart axes. Mirrors the web viewer's
//! `assets/lib/charts/util/units.js` so the terminal shows the same scaled
//! units (e.g. `100.5%`, `503.32ms`, `9.29GB`, `10.34Gbps`, `47.9K/s`).
//!
//! Each entry is `(threshold, suffix, divisor, multiplier)`: pick the
//! largest scale whose `threshold` the absolute value reaches, then render
//! `value * multiplier / divisor` with the suffix.

type Scale = (f64, &'static str, f64, f64);

// Time, base nanoseconds.
static TIME: &[Scale] = &[
    (0.0, "ns", 1.0, 1.0),
    (1e3, "μs", 1e3, 1.0),
    (1e6, "ms", 1e6, 1.0),
    (1e9, "s", 1e9, 1.0),
];

// Data size, base bytes (binary scaling, matching the web viewer).
static BYTES: &[Scale] = &[
    (0.0, "B", 1.0, 1.0),
    (1024.0, "KB", 1024.0, 1.0),
    (1_048_576.0, "MB", 1_048_576.0, 1.0),
    (1_073_741_824.0, "GB", 1_073_741_824.0, 1.0),
    (1_099_511_627_776.0, "TB", 1_099_511_627_776.0, 1.0),
];

// Network bit rate, base bits/sec.
static BITRATE: &[Scale] = &[
    (0.0, "bps", 1.0, 1.0),
    (1e3, "Kbps", 1e3, 1.0),
    (1e6, "Mbps", 1e6, 1.0),
    (1e9, "Gbps", 1e9, 1.0),
    (1e12, "Tbps", 1e12, 1.0),
];

// Data rate, base bytes/sec.
static DATARATE: &[Scale] = &[
    (0.0, "B/s", 1.0, 1.0),
    (1e3, "KB/s", 1e3, 1.0),
    (1e6, "MB/s", 1e6, 1.0),
    (1e9, "GB/s", 1e9, 1.0),
    (1e12, "TB/s", 1e12, 1.0),
];

// Percentage: value is a fraction, scaled by 100.
static PERCENTAGE: &[Scale] = &[(0.0, "%", 1.0, 100.0)];

// Frequency, base Hz.
static FREQUENCY: &[Scale] = &[
    (0.0, "Hz", 1.0, 1.0),
    (1e3, "KHz", 1e3, 1.0),
    (1e6, "MHz", 1e6, 1.0),
    (1e9, "GHz", 1e9, 1.0),
];

// Bare count with K/M/B suffixes.
static COUNT: &[Scale] = &[
    (0.0, "", 1.0, 1.0),
    (1e3, "K", 1e3, 1.0),
    (1e6, "M", 1e6, 1.0),
    (1e9, "B", 1e9, 1.0),
];

// Rate (count per second) with /s, K/s, M/s, B/s suffixes.
static RATE: &[Scale] = &[
    (0.0, "/s", 1.0, 1.0),
    (1e3, "K/s", 1e3, 1.0),
    (1e6, "M/s", 1e6, 1.0),
    (1e9, "B/s", 1e9, 1.0),
];

fn scales(unit_system: &str) -> Option<&'static [Scale]> {
    match unit_system {
        "time" => Some(TIME),
        "bytes" => Some(BYTES),
        "bitrate" => Some(BITRATE),
        "datarate" => Some(DATARATE),
        "percentage" => Some(PERCENTAGE),
        "frequency" => Some(FREQUENCY),
        "count" => Some(COUNT),
        "rate" => Some(RATE),
        _ => None,
    }
}

/// Format `value` (in the unit system's base unit) with auto-scaling and a
/// suffix. An unknown or absent unit system falls back to a plain, trimmed
/// number. Two decimals, trailing zeros removed (`100.50%` -> `100.5%`,
/// `9.00` -> `9`).
pub fn format_value(unit_system: Option<&str>, value: f64) -> String {
    if !value.is_finite() {
        return "—".to_string();
    }
    let Some(sys) = unit_system.and_then(scales) else {
        return trim(&format!("{value:.2}")).to_string();
    };
    let av = value.abs();
    let &(_, suffix, divisor, multiplier) = sys
        .iter()
        .rev()
        .find(|&&(th, ..)| av >= th)
        .unwrap_or(&sys[0]);
    let scaled = value * multiplier / divisor;
    format!("{}{}", trim(&format!("{scaled:.2}")), suffix)
}

/// Trim trailing fractional zeros (and a bare trailing dot) from a decimal
/// string. Leaves integer strings untouched.
fn trim(s: &str) -> &str {
    if s.contains('.') {
        s.trim_end_matches('0').trim_end_matches('.')
    } else {
        s
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn percentage_scales_by_100() {
        assert_eq!(format_value(Some("percentage"), 1.005), "100.5%");
        assert_eq!(format_value(Some("percentage"), 0.152), "15.2%");
        assert_eq!(format_value(Some("percentage"), 0.0), "0%");
    }

    #[test]
    fn time_scales_from_nanoseconds() {
        // 1.535 is 1.5349…9 in f64, so `{:.2}` yields 1.53 — matching the
        // web viewer's toFixed(2) on the same value.
        assert_eq!(format_value(Some("time"), 1535.0), "1.53μs");
        assert_eq!(format_value(Some("time"), 503_316_479.0), "503.32ms");
        assert_eq!(format_value(Some("time"), 2_000_000_000.0), "2s");
        assert_eq!(format_value(Some("time"), 42.0), "42ns");
    }

    #[test]
    fn bytes_use_binary_scaling() {
        assert_eq!(format_value(Some("bytes"), 9_980_456_960.0), "9.3GB");
        assert_eq!(format_value(Some("bytes"), 1024.0), "1KB");
    }

    #[test]
    fn bitrate_uses_si_scaling() {
        assert_eq!(format_value(Some("bitrate"), 10_338_674_776.0), "10.34Gbps");
    }

    #[test]
    fn rate_and_count_suffixes() {
        assert_eq!(format_value(Some("rate"), 47902.0), "47.9K/s");
        assert_eq!(format_value(Some("rate"), 500.0), "500/s");
        assert_eq!(format_value(Some("count"), 1_500_000.0), "1.5M");
    }

    #[test]
    fn unknown_or_missing_unit_is_plain_trimmed() {
        assert_eq!(format_value(None, 42.5), "42.5");
        assert_eq!(format_value(None, 9.0), "9");
        assert_eq!(format_value(Some("bogus"), 123.456), "123.46");
    }
}
